File: 0210-course-schedule-ii/0210-course-schedule-ii.cpp
class Solution {
public:
bool checkCycle( int node ,vector<vector<int>> &graph, vector<int> & vis ,vector<int>&visDfs )
{
    vis[node] =1;
    visDfs[node] =1;

    for( int  i= 0 ;i< graph[node].size() ;i++ )
    {
        if( !vis[graph[node][i]])
        {
            checkCycle(graph[node][i], graph, vis, visDfs);
        }
        if(vis[graph[node][i]] && visDfs[graph[node][i]])
            return true;
    }
    visDfs[node]=0 ;
    return false;
}

bool isCycle (vector<vector<int>> &graph , int n)
{
    vector<int>vis(n ,0 );
    vector<int>visDfs(n,0);
    for( int i=0 ; i< n; i++ )
    {
        if( !vis[i])
        {
            if(checkCycle(i,graph,vis,visDfs))
                return true;
        }
    }
    return false;
}

void TopologicaalSort(  int u ,vector<vector<int>>&graph , vector<int> &vis ,stack<int>&st)
{
    vis[u] =1;
    for( int i= 0 ;i< graph[u].size() ;i++ )
    {
        if( !vis[graph[u][i]])
        {
            TopologicaalSort(graph[u][i], graph,vis ,st);
        }
    }
    st.push(u);
}
vector<int> TopologicalSortUnit (vector<vector<int>>&graph ,int num )
{
    vector<int>v;
    stack<int> st;
    vector<int> vis(num , 0);
    for( int i =0 ;i< graph.size() ;i++ )
    {
        if(!vis[i])
            TopologicaalSort(i,graph,vis,st);
    }
    while(!st.empty())
    {
        v.push_back(st.top());
        st.pop();
    }
    return v;
}

vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites)
{
    vector<vector<int> > graph (numCourses );
    for( int i=0 ;i< prerequisites.size() ;i++ )
    {
        graph[prerequisites[i][1]].push_back(prerequisites[i][0]);   
    }
    if(isCycle(graph,numCourses))
    {
        return {};
    }
    vector<int>ans = TopologicalSortUnit(graph ,numCourses);
    return ans;
}
};

File: 0210-course-schedule-ii/0210-course-schedule-ii.js
var isCyclicMap = function (adj, vis, finish, node) {
  vis.set(node, 1);
  let cyclic = 0;
  let neighpors = adj.get(node) || [];
  for (let i = 0; i < neighpors.length; i++) {
    let neighbor = neighpors[i];
    if (!vis.has(neighbor))
      cyclic = cyclic || isCyclicMap(adj, vis, finish, neighbor);
    else if (finish.get(neighbor) === 0) cyclic = 1;
  }
  finish.set(node, 1);
  return cyclic;
};

var dfs = function (node, graph, visited, courses) {
  visited.set(node, 1);
  let neighpors = graph.get(node) || [];
  for (let i = 0; i < neighpors.length; i++) {
    let neighpor = neighpors[i];

    if (!visited.has(neighpor)) {
      dfs(neighpor, graph, visited, courses);
    }
  }
  courses.push(node);
};

var findOrder = function (numCourses, prerequisites) {
  let graph = new Map();
  let visited = new Map();
  let courses = [];
  let finish = new Map();

  for (let i = 0; i < numCourses; i++) {
    finish.set(i, 0);
  }

  for (let i = 0; i < prerequisites.length; i++) {
    finish.set(i, 0);
    let [c1, c2] = prerequisites[i];
    if (!graph.has(c1)) {
      graph.set(c1, [c2]);
    } else {
      graph.get(c1).push(c2);
    }
  }

  for (let i = 0; i < numCourses; i++) {
    if (isCyclicMap(graph, visited, finish, i)) {
      return [];
    }
  }

  visited.clear();

  for (let i = 0; i < numCourses; i++) {
    if (!visited.has(i)) {
      dfs(i, graph, visited, courses);
    }
  }
  return courses;
};
